### PR TITLE
BMLMI Implementation

### LINKED ISSUE #233: Implementation of method type="BMLMI" - bootstrapped ML MI as in van Hippel & Bartlett
Method is described in van Hippel & Bartlett, Stat Sci 2021 article also available at [https://arxiv.org/pdf/1210.0870.pdf](https://arxiv.org/pdf/1210.0870.pdf).
The methods is based on analyzing D randomly imputed datasets (imputation uses the ML/REML estimator) for each of B bootstrap samples. Estimates are then pooled across the B*D datasets and variance is estimated as per section 8 of the manuscript. 

This is less efficient than using conditional mean imputation for ANCOVA as the analysis method but has the advantage to also be theoretically valid (in the sense of giving correct frequentist inference) for other analysis models.

### PR DESCRIPTION
**Pull request by nociale**
Closes #233

### CODE REVIEW

**Comment by nociale** [comment]
Addresses #233

**Comment by gowerc** [comment]
Hey @nociale , Just to say that you now have a merge conflict that needs to be resolved..

File: R/draws.R [hunk 1]
```diff
@@ -50,10 +50,14 @@
 #' MMRM parameter estimates from the original dataset and from each leave-one-subject-out sample.
 #' This method can be specified by using `method = method_condmean()` with
 #' argument `type = "jackknife"`.
+#' - Bootstrapped Maximum Likelihood MI: `draws` returns the MMRM parameter estimates from
+#' a given number of bootstrap samples needed to perform random imputations of the bootstrapped samples.
+#' This method can be specified by using `method = method_bmlmi()`.
```
Comment (gowerc) [inline R/draws.R:55]: This description doesn't feel that clear to me (maybe from the double use of "bootstrapped").  Will try and re-word later but I think we should change it somehow

File: R/draws.R [hunk 2]
```diff
@@ -152,6 +156,9 @@
 #' Marcel Wolbers, Alessandro Noci, Paul Delmar, Craig Gower-Page, Sean Yiu, Jonathan W. Bartlett. Reference-based
 #' imputation methods based on conditional mean imputation. \url{http://arxiv.org/abs/2109.11162}, 2021.
 #'
+#' Von Hippel, Paul T and Bartlett, Jonathan W.
+#' Maximum likelihood multiple imputation: Faster imputations and consistent standard errors without posterior draws. 2021.
+#'
```
Comment (gowerc) [inline R/draws.R:161]: Do we have any document id numbers or a url ? (just thinking that we've provided number references in all of our above references ? )

File: R/impute.R [hunk 3]
```diff
@@ -158,6 +163,8 @@ impute_internal <- function(draws, references, update_strategy, strategies, cond
     validate(draws$samples)
     samples_grouped <- transpose_samples(draws$samples)
 
+    n_imputations = ifelse(is.null(draws$method$D), 1, draws$method$D)
```
Comment (gowerc) [inline R/impute.R:166]:
```suggestion
    n_imputations <- ifelse(is.null(draws$method$D), 1, draws$method$D)
```

File: R/impute.R [hunk 4]
```diff
@@ -432,7 +460,7 @@ impute_data_individual <- function(
     if (condmean) {
         imputed_outcome <- lapply(conditional_parameters, function(x) as.vector(x$mu))
     } else {
-        imputed_outcome <- lapply(conditional_parameters, impute_outcome)
+        imputed_outcome <- lapply(conditional_parameters, function(x) impute_outcome(x, n_imputations))
```
Comment (gowerc) [inline R/impute.R:463]:
```suggestion
        imputed_outcome <- lapply(conditional_parameters, impute_outcome, n_imputations = n_imputations)
```

**Comment by gowerc** [comment]
@nociale  looks like the B parameter in method_bmlmi is being ignored / replaced 

```
> drawobj <- draws(
+     data = dat,
+     data_ice = dat_ice,
+     vars = vars,
+     method = method_bmlmi(B = 2, D = 4)
+ )
drawobj
> drawobj

Draws Object
------------
Number of Samples: 20
Number of Failed Samples: 0
Model Formula: outcome ~ 1 + group + visit + age + sex + visit * group
Imputation Type: random
Method:
    covariance: us
    threshold: 0.01
    same_cov: TRUE
    REML: TRUE
    B: 20
    D: 2
    n_samples: NULL

```

File: R/impute.R [hunk 5]
```diff
@@ -486,21 +514,32 @@ get_visit_distribution_parameters <- function(dat, beta, sigma) {
...
+    result <- do.call(
+        rbind,
+        lapply(
+            seq.int(n_imputations),
+            function(x)
+            sample_mvnorm(
+                conditional_parameters$mu,
+                conditional_parameters$sigma
+            )
+        )
```
Comment (gowerc) [inline R/impute.R:539]:
```suggestion
    samps <- replicate(
        n = n_imputations,
        simplify = FALSE,
        expr = {
            sample_mvnorm(
                conditional_parameters$mu,
                conditional_parameters$sigma
            )
        }
    )
    result <- matrix(nrow = n_imputations, unlist(samps))
```

**Comment by nociale** [comment]
@gowerc Thanks for highlighting the problem. I have found the issue (it was just that `method_bmlmi` was not updating the default values)

File: R/draws.R (hunk 1 above)
Comment (nociale) [reply R/draws.R:55]: mmm, I agree but at the same time if I think of it I kind of understand that I need to specify twice "bootstrapped samples".. I am happy to change it however.

File: R/draws.R (hunk 2 above)
Comment (nociale) [reply R/draws.R:161]: Sorry, what this the difference between this reference and the other ones? Some of them have a url, but only those ones that are in arXiv I think. I can add more details like Journal etc...?

File: R/impute.R [hunk 6]
```diff
@@ -486,21 +514,33 @@ get_visit_distribution_parameters <- function(dat, beta, sigma) {
...
+        simplify = FALSE,
+        expr = {
+            sample_mvnorm(
+                conditional_parameters$mu,
+                conditional_parameters$sigma
+            )
+        }
+    )
+    result <- matrix(nrow = n_imputations, unlist(samps))
     )
```
Comment (nociale) [inline R/impute.R:541]:
```suggestion

```

**Comment by nociale** [comment]
@gowerc It seems that we can merge it to develop if you agree!

**Review by gowerc — APPROVED** [review]

**Comment by gowerc** [comment]
@nociale ,  Feel free to merge this in though please don't do so until #276  has been merged

**Pull request merged**